presto-main/src/main/java/com/facebook/presto/operator/ContinuousWorkUtils.java

```diff
@@ -0,0 +1,322 @@
+                    return WorkState.ofResult(elementAndWork.getElement());
+                }
+            }
+        });
+    }
+
+    static <T, R> ContinuousWork<R> flatMap(ContinuousWork<T> work, Function<T, Iterator<R>> mapper)
+    {
+        return flatten(map(work, mapper));
+    }
+
+    private static <T> ContinuousWork<T> flatten(ContinuousWork<Iterator<T>> work)
+    {
+        return work.transform(iteratorOptional ->
+                iteratorOptional
+                        .map(iterator -> {
+                            if (iterator.hasNext()) {
+                                return WorkState.ofResult(iterator.next(), false);
+                            }
+                            else {
+                                return WorkState.<T>needsMoreData();
+                            }
+                        })
+                        .orElse(WorkState.finished()));
+    }
+
+    static <T, R> ContinuousWork<R> map(ContinuousWork<T> work, Function<T, R> mapper)
+    {
+        return work.transform(elementOptional ->
+                elementOptional
+                        .map(element -> WorkState.ofResult(mapper.apply(element)))
+                        .orElse(WorkState.finished()));
+    }
+
+    static <T, R> ContinuousWork<R> transform(ContinuousWork<T> work, Function<Optional<T>, WorkState<R>> transformation)
+    {
+        return create(new Supplier<WorkState<R>>()
+        {
+            Optional<T> element = Optional.empty();
+
+            @Override
+            public WorkState<R> get()
+            {
+                while (true) {
+                    if (!element.isPresent() && !work.isFinished()) {
+                        if (work.process()) {
+                            if (!work.isFinished()) {
+                                element = Optional.of(work.getResult());
+                            }
+                        }
+                        else if (work.isBlocked()) {
+                            return WorkState.blocked(work.getBlockedFuture());
+                        }
+                        else {
+                            return WorkState.yield();
+                        }
+                    }
+
+                    WorkState<R> state = requireNonNull(transformation.apply(element), "state is null");
+
+                    if (state.needsMoreData) {
+                        checkState(!work.isFinished(), "Cannot request more data when base work is finished");
+                        // set element to empty() in order to fetch a new one
+                        element = Optional.empty();
+                    }
+
+                    if (state.type != WorkState.Type.NEEDS_MORE_DATA) {
+                        // passthrough transformation state if it doesn't require new data
+                        return state;
+                    }
+                }
+            }
+        });
+    }
+
+    static <T> ContinuousWork<T> create(Supplier<WorkState<T>> supplier)
```

Comment:
Supplier is not good interface here. You are calling it multiple times assuming it is doing work. Please provide different functional interface (`Processor` or whatever).